### DISCUSSION TITLE
ParallelProcessor usage

### DISCUSSION THREAD

**Discussion by keithchan9999** [CPJKU/madmom#516] [General]
### Expected behaviour

Please describe how you expected madmom to behave.

### Actual behaviour

I am trying to run both a beat processor and a tempo processor which wraps under a parallel processor.  But I got an error like this: 

  File "mad.py", line 158, in main
    madmom.processors.process_online(parallel_processor, **kwargs)
  File "C:\Users\kido9\miniconda3\envs\mad\lib\site-packages\madmom\processors.py", line 885, in process_online
    _process((processor, frame, outfile, process_args))
  File "C:\Users\kido9\miniconda3\envs\mad\lib\site-packages\madmom\processors.py", line 289, in _process
    return process_tuple[0](*process_tuple[1:-1], **process_tuple[-1])
  File "C:\Users\kido9\miniconda3\envs\mad\lib\site-packages\madmom\processors.py", line 120, in __call__
    return self.process(*args, **kwargs)
TypeError: process() takes 2 positional arguments but 3 were given

Tried to troubleshoot but unfortunately I ain't any expert on that.
 
### Steps needed to reproduce the behaviour

```python
kwargs = dict(
		fps = 100,
		correct = True,
		infile = None,
		outfile = None,
		max_bpm = 150,
		min_bpm = 60,
		num_frames = 1,
		online = True,
		stream_input_device = dev_index,
		#nn_files = [BEATS_LSTM[0], BEATS_LSTM[1], BEATS_LSTM[2], BEATS_LSTM[3], BEATS_LSTM[4]]
		#verbose = 1
	)

	in_processor = madmom.features.beats.RNNBeatProcessor(**kwargs)
	beat_processor = madmom.features.beats.DBNBeatTrackingProcessor(**kwargs)
	out_processor = [beat_processor, beat_callback]
	processor = madmom.processors.IOProcessor(in_processor, out_processor)

	tempo_processor = madmom.features.tempo.TempoEstimationProcessor(**kwargs)
	out2_processor = [tempo_processor, tempo_callback]
	processor2 = madmom.processors.IOProcessor(in_processor, out2_processor)

	parallel_processor = madmom.processors.ParallelProcessor([processor, processor2], num_threads=2)
	madmom.processors.process_online(parallel_processor, **kwargs)

```

### Information about installed software

Please provide some information about installed software.

```python
import madmom
import numpy as np
import scipy

madmom.__version__
np.__version__
scipy.__version__
```

**Comment by superbock** [comment]
madmom does not have such a sophisticated process management, so you need to do this by yourself, i.e. wrap the beat tempo sub-pipelines in a parallel processor and use in_processor to feed it.

```
in_processor = madmom.features.beats.RNNBeatProcessor(**kwargs)

beat_tracker = madmom.features.beats.DBNBeatTrackingProcessor(**kwargs)
beat_processor = madmom.processors.IOProcessor(beat_tracker, beat_callback)

tempo_estimator = madmom.features.tempo.TempoEstimationProcessor(**kwargs)
tempo_processor = madmom.processors.IOProcessor(tempo_estimator, tempo_callback)

parallel_processor = madmom.processors.ParallelProcessor([beat_processor, tempo_processor], num_threads=2)

pipeline = madmom.processors.SequentialProcessor([in_processor, parallel_processor])
madmom.processors.process_online(pipeline, **kwargs)
```

However, you have to make sure the your `beat_callback` and `tempo_callback` accept two arguments `(data, output)`.

HTH,

**Reply by keithchan9999** [reply]
Thanks for your reply,  I think I've tried something very similar.  But the above code is suffering the same problem as before:

```python
  File "mad.py", line 163, in main
    pipeline = madmom.processors.SequentialProcessor(in_processor, parallel_processor)
TypeError: __init__() takes 2 positional arguments but 3 were given
```

**Reply by superbock** [reply]
Sorry, my fault, of course it has to be `pipeline = madmom.processors.SequentialProcessor([in_processor, parallel_processor])` [I edited the comment above to reflect this].

**Reply by keithchan9999** [reply]
Errrrrrr.... I am sorry that it still give the same error.

    madmom.processors.process_online(pipeline, **kwargs)
  File "C:\Users\kido9\miniconda3\envs\mad\lib\site-packages\madmom\processors.py", line 885, in process_online
    _process((processor, frame, outfile, process_args))
  File "C:\Users\kido9\miniconda3\envs\mad\lib\site-packages\madmom\processors.py", line 289, in _process
    return process_tuple[0](*process_tuple[1:-1], **process_tuple[-1])
  File "C:\Users\kido9\miniconda3\envs\mad\lib\site-packages\madmom\processors.py", line 120, in __call__
    return self.process(*args, **kwargs)
TypeError: process() takes 2 positional arguments but 3 were given

**Reply by superbock** [reply]
Sorry, for not testing this before posting. You have to put the output processors after the parallel processing of beats and tempo. Here's a working version:

```
def write_outputs(data, outfile):
    madmom.io.write_beats(data[0], outfile + '.beats')
    madmom.io.write_tempo(data[1], outfile + '.tempo')


in_processor = madmom.features.beats.RNNBeatProcessor(**kwargs)
beat_processor = madmom.features.beats.DBNBeatTrackingProcessor(**kwargs)
tempo_processor = madmom.features.tempo.TempoEstimationProcessor(**kwargs)
parallel_processor = madmom.processors.ParallelProcessor([beat_processor, tempo_processor], num_threads=2)
pipeline = madmom.processors.SequentialProcessor([in_processor, parallel_processor])
out_processor = madmom.processors.IOProcessor(pipeline, write_outputs)
madmom.processors.process_single(out_processor, **kwargs)
```

**Reply by keithchan9999** [reply]
thanks for your reply.  There are still errors on that.  It looked like the 'process_single' treat the input from a file. It has the following errors, and also I tried to replace `process_single` with `process_online` which wasn't success either.

```python
   data = _process((processor, data, kwargs))
  File "C:\Users\KeithChan\anaconda3\envs\mad\lib\site-packages\madmom\processors.py", line 289, in _process
    return process_tuple[0](*process_tuple[1:-1], **process_tuple[-1])
  File "C:\Users\KeithChan\anaconda3\envs\mad\lib\site-packages\madmom\processors.py", line 120, in __call__
    return self.process(*args, **kwargs)
  File "C:\Users\KeithChan\anaconda3\envs\mad\lib\site-packages\madmom\processors.py", line 424, in process
    data = _process((processor, data, kwargs))
  File "C:\Users\KeithChan\anaconda3\envs\mad\lib\site-packages\madmom\processors.py", line 289, in _process
    return process_tuple[0](*process_tuple[1:-1], **process_tuple[-1])
  File "C:\Users\KeithChan\anaconda3\envs\mad\lib\site-packages\madmom\processors.py", line 120, in __call__
    return self.process(*args, **kwargs)
  File "C:\Users\KeithChan\anaconda3\envs\mad\lib\site-packages\madmom\processors.py", line 424, in process
    data = _process((processor, data, kwargs))
  File "C:\Users\KeithChan\anaconda3\envs\mad\lib\site-packages\madmom\processors.py", line 289, in _process
    return process_tuple[0](*process_tuple[1:-1], **process_tuple[-1])
  File "C:\Users\KeithChan\anaconda3\envs\mad\lib\site-packages\madmom\processors.py", line 120, in __call__
    return self.process(*args, **kwargs)
  File "C:\Users\KeithChan\anaconda3\envs\mad\lib\site-packages\madmom\audio\signal.py", line 769, in process
    return Signal(data, **args)
  File "C:\Users\KeithChan\anaconda3\envs\mad\lib\site-packages\madmom\audio\signal.py", line 598, in __new__
    dtype=dtype)
  File "C:\Users\KeithChan\anaconda3\envs\mad\lib\site-packages\madmom\io\audio.py", line 605, in load_audio_file
    stop=stop, dtype=dtype)
  File "C:\Users\KeithChan\anaconda3\envs\mad\lib\site-packages\madmom\io\audio.py", line 493, in load_wave_file
    file_sample_rate, signal = wavfile.read(filename, mmap=True)
  File "C:\Users\KeithChan\anaconda3\envs\mad\lib\site-packages\scipy\io\wavfile.py", line 647, in read
    fid = open(filename, 'rb')
TypeError: expected str, bytes or os.PathLike object, not NoneType
```

**Reply by keithchan9999** [reply]
Actually, I was aiming to have that online, ie feed from the mic, so that's probably why it wouldn't work......  Also, I would like to learn more about how to use the Processor, it would be nicer if more examples could be provided, do you know anywhere I could read more on this?

**Reply by superbock** [reply]
Of course you have to supply a valid input file for the above example to work. You can also replace `process_single` with `process_online` of course if everything else is working.

Good examples on how to use the processors are the programs provided in the `/bin` directory.

**Reply by keithchan9999** [reply]
```python
in_processor = madmom.features.beats.RNNBeatProcessor(**kwargs)
beat_processor = madmom.features.beats.DBNBeatTrackingProcessor(**kwargs)
tempo_processor = madmom.features.tempo.TempoEstimationProcessor(**kwargs)
parallel_processor = madmom.processors.ParallelProcessor([beat_processor, tempo_processor], num_threads=2)
pipeline = madmom.processors.SequentialProcessor([in_processor, parallel_processor])
out_processor = madmom.processors.IOProcessor(pipeline, write_outputs)
madmom.processors.process_single(out_processor, **kwargs)
```

Strangely it didn't output anything if I have [beat_processor, tempo_processor] when constructing ParallelProcessor as above, however, when I just use [beat_processor], it was able to output the correct beat information..... Do you know if is that a known bug?

**Reply by superbock** [reply]
Because with your settings, you're processing only a single frame and stop processing afterwards. Try unsetting `num_frames` in your `kwargs`.
